### DISCUSSION TITLE
Adding variable on OPC-UA Server writeable from Clients

### DISCUSSION THREAD

**Discussion by gteti** [FreeOpcUa/opcua-asyncio#1086] [Q&A] [answered]
Hello everyone,
I'm trying to create a writeable variable on OPC-UA Server. I'm using the standard server example https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/server-example.py where I first added a variable like:

```python
myvar2 = await myobj.add_variable(idx, "myvar", 0.0)
await myvar2.set_writable()
```

And after not being able to change it, I tried adding a new object and set that object to writable as:
```python
myobj2 = await server.nodes.objects.add_object(idx, "MyObject2")
await myobj2.set_writable()
myvar2 = await myobj2.add_variable(idx, "myvar", 0.0)
await myvar2.set_writable()
```

But I'm still not able to write any value to it from a client. In this last configuration I even get an error:
```bash
Traceback (most recent call last):
  File "C:\InLav\Gianluca\Python\Server_opcua\test.py", line 153, in <module>
    asyncio.run(main())
  File "C:\Anaconda3\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Anaconda3\lib\asyncio\base_events.py", line 647, in run_until_complete
    return future.result()
  File "C:\InLav\Gianluca\Python\Server_opcua\test.py", line 106, in main
    await myobj2.set_writable()
  File "C:\Anaconda3\lib\site-packages\asyncua\common\node.py", line 244, in set_writable
    await self.set_attr_bit(ua.AttributeIds.AccessLevel, ua.AccessLevel.CurrentWrite)
  File "C:\Anaconda3\lib\site-packages\asyncua\common\node.py", line 251, in set_attr_bit
    dv = await self.read_attribute(attr)
  File "C:\Anaconda3\lib\site-packages\asyncua\common\node.py", line 304, in read_attribute
    result[0].StatusCode.check()
  File "C:\Anaconda3\lib\site-packages\asyncua\ua\uatypes.py", line 328, in check
    raise UaStatusCodeError(self.value)
asyncua.ua.uaerrors._auto.BadAttributeIdInvalid: "The attribute is not supported for the specified Node."(BadAttributeIdInvalid)
```

The error happens when I use the `await myobj2.set_writable()`.

Can you help me ? Am I doing something wrong ? 

Thank you

**Comment by AndreasHeine** [comment]
what does the client return? an ErrorStatusCode like BadNotWritable?
how dose the AttributesTab with UAExpert look like?

**Comment by gteti** [comment]
Sorry to have bothered you. I've used UaExpert instead of a python opcua-client and was able to write the value to "myvar" correctly. Then I searched what was wrong with the other client and all seems to work.

I get this error when I use the `await myobj2.set_writable()`:
```bash
Traceback (most recent call last):
  File "C:\InLav\Gianluca\Python\Server_opcua\test.py", line 153, in <module>
    asyncio.run(main())
  File "C:\Anaconda3\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Anaconda3\lib\asyncio\base_events.py", line 647, in run_until_complete
    return future.result()
  File "C:\InLav\Gianluca\Python\Server_opcua\test.py", line 106, in main
    await myobj2.set_writable()
  File "C:\Anaconda3\lib\site-packages\asyncua\common\node.py", line 244, in set_writable
    await self.set_attr_bit(ua.AttributeIds.AccessLevel, ua.AccessLevel.CurrentWrite)
  File "C:\Anaconda3\lib\site-packages\asyncua\common\node.py", line 251, in set_attr_bit
    dv = await self.read_attribute(attr)
  File "C:\Anaconda3\lib\site-packages\asyncua\common\node.py", line 304, in read_attribute
    result[0].StatusCode.check()
  File "C:\Anaconda3\lib\site-packages\asyncua\ua\uatypes.py", line 328, in check
    raise UaStatusCodeError(self.value)
asyncua.ua.uaerrors._auto.BadAttributeIdInvalid: "The attribute is not supported for the specified Node."(BadAttributeIdInvalid)
```

**Reply by AndreasHeine** [reply]
objects aren't writable at all only variables an properties with the attribute value!

**Comment by gteti** [comment, marked as answer]
**Answer** explained by [AndreasHeine](https://github.com/AndreasHeine) in the comments!

Is there a way to allocate a variable at a specific address ?

I have a project which uses the old version of opcua server. So I'd like for something like

ns=2;i=30 for my variable (with add_variable)

**Reply by AndreasHeine** [reply]
for the Identifier (30 in your example) the answer is yes!
for the NamespaceIndex (2 in your example) it depends...

the NamespaceIndex is the index of the NameSpcaeUri in the servers NamespaceArray so if the ns=2 is already occupied it will get the next free index... 

usually for mappings we use the ExpandedNodeId-String annotation "nsu=http://mynamespace;i=30"
and after connecting to the server we read the NamespaceArray figgure out which NamespaceIndex my ExpandedNodeId has in the Server and then i use the Index and the Identifier to build the correct NodeId

same on server-side!

the NamespaceIndex is not fix by definition but the NamespaceUri is!

**Reply by gteti** [reply]
I've added the 'nsu' instead of 'ns' as shown in the example and it worked. I was able to create the object using the mynamespace created with *register_namespace* the problem is that the i=30 doesn't produce a variable allocated at that address.

I've used this example: https://github.com/FreeOpcUa/python-opcua/blob/master/examples/server-custom-object.py

**Reply by AndreasHeine** [reply]
you need to understand the namespaceconcept of opcua
ns=0 -> OPC UA Core Namespace
ns=1 -> OPC UA Server own namespace
ns=2 -> custom registered namespace or companionspec. namespace (xml imported namespace)
ns=3 -> ...
ns=4 -> "http://mynamespace"
ns=5 -> ...

Example:

"nsu=http://mynamespace;i=30/"

1. read namespace array -> [... , ... , ... , ... , "http://mynamespace", ...]
2. get index of my nsu -> idx=4 (identifier=30)
3. create a nodeid f"ns={idx};i={identifier}"
4. use class `NodeId.from_string(f"ns={idx};i={identifier}")` to get your nodeid instance
5. create a node using the nodeid

**Reply by gteti** [reply]
Thank you for the explaination. 
I am trying to do exactly that:
```python
opcua_idx = opcua_server.register_namespace(uri)

myobjtest_nodeid = ua.NodeId.from_string('ns=%d;i=40' % opcua_idx)  
        
myobjtest = objects.add_object(opcua_idx, "Myobj", myobjtest_nodeid)

myvartest_str = f"myvartest"
myvartest = myobjtest.add_variable(opcua_idx , myvartest_str, 0)
```

This created object is not in the `i=30` position and allocated variable under it get also a different *_*.

In your example using a `ns=4` will create the object at `ns=4;i=1` (even if I add `i=40`) and the variable added to it will be `ns=4;i=2`.

**Reply by AndreasHeine** [reply]
![image](https://user-images.githubusercontent.com/56362817/196354587-181c8141-ce60-4b26-9256-82c175dd7c26.png)

![image](https://user-images.githubusercontent.com/56362817/196354504-8e098e56-233f-4eea-932a-74110bdafc0c.png)

```python
    opcua_idx = await server.register_namespace("http://test")
    myobjtest_nodeid = ua.NodeId.from_string('ns=%d;i=40' % opcua_idx)  
    myobjtest = await server.nodes.objects.add_object(myobjtest_nodeid, "Myobj")
    myvartest_str = f"myvartest"
    myvartest = await myobjtest.add_variable(
        ua.NodeId.from_string('ns=%d;i=30' % opcua_idx), 
        myvartest_str,
        0.0,
        ua.VariantType.Float
    )
```

**Reply by AndreasHeine** [reply]
if you pass the index as first arg like you did it will generate a id -> if you pass the nodeid-class it use that instead ;)

**Reply by gteti** [reply]
Thank you, it works!
I was following the examples and the creation of the object was not starting with the nodeid and the variable too. 

🥇
